<a href="https://colab.research.google.com/github/claudiocdc/fifa-world-cup-analysis/blob/main/01_Limpieza_Fifa_WorldCup.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/claudiocdc/fifa-world-cup-analysis/main/WorldCups.csv"
df = pd.read_csv(url)
print(df.head())
print(df.info())

   Year      Country      Winner      Runners-Up    Third      Fourth  \
0  1930      Uruguay     Uruguay       Argentina      USA  Yugoslavia   
1  1934        Italy       Italy  Czechoslovakia  Germany     Austria   
2  1938       France       Italy         Hungary   Brazil      Sweden   
3  1950       Brazil     Uruguay          Brazil   Sweden       Spain   
4  1954  Switzerland  Germany FR         Hungary  Austria     Uruguay   

   GoalsScored  QualifiedTeams  MatchesPlayed Attendance  
0           70              13             18    590.549  
1           70              16             17    363.000  
2           84              15             18    375.700  
3           88              13             22  1.045.246  
4          140              16             26    768.607  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Year       

In [3]:
#vemos que la asistencia son texto porque tienen el punto, la limpiamos para pasarlo a número


df['Attendance'] = df['Attendance'].str.replace(".","")
print(df.head())

   Year      Country      Winner      Runners-Up    Third      Fourth  \
0  1930      Uruguay     Uruguay       Argentina      USA  Yugoslavia   
1  1934        Italy       Italy  Czechoslovakia  Germany     Austria   
2  1938       France       Italy         Hungary   Brazil      Sweden   
3  1950       Brazil     Uruguay          Brazil   Sweden       Spain   
4  1954  Switzerland  Germany FR         Hungary  Austria     Uruguay   

   GoalsScored  QualifiedTeams  MatchesPlayed Attendance  
0           70              13             18     590549  
1           70              16             17     363000  
2           84              15             18     375700  
3           88              13             22    1045246  
4          140              16             26     768607  


In [4]:
#cambiamos a número entero int
df['Attendance'] = df['Attendance'].astype(int)
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Year            22 non-null     int64 
 1   Country         22 non-null     object
 2   Winner          22 non-null     object
 3   Runners-Up      22 non-null     object
 4   Third           22 non-null     object
 5   Fourth          22 non-null     object
 6   GoalsScored     22 non-null     int64 
 7   QualifiedTeams  22 non-null     int64 
 8   MatchesPlayed   22 non-null     int64 
 9   Attendance      22 non-null     int64 
dtypes: int64(5), object(5)
memory usage: 1.8+ KB
None


**¿Qué mundial lleno más estadios?**

In [5]:
df.sort_values('Attendance',ascending = False).head()

,Year,Country,Winner,Runners-Up,Third,Fourth,GoalsScored,QualifiedTeams,MatchesPlayed,Attendance
14,1994,USA,Brazil,Italy,Sweden,Bulgaria,141,24,52,3587538
21,2022,Qatar,Argentina,France,Croatia,Morocco,172,32,64,3404252
19,2014,Brazil,Germany,Argentina,Netherlands,Brazil,171,32,64,3386810
17,2006,Germany,Italy,France,Germany,Portugal,147,32,64,3359439
18,2010,South Africa,Spain,Netherlands,Germany,Uruguay,145,32,64,3178856


El Mundial con mayor asistencia total de la historia fue USA 1994, con más de 3,58 millones de espectadores.

In [6]:
#sorprendente que con menos partidos hubiese mas asistencia que el de 2014 que tuvo mas partidos, suponemos que es por los estadios americanos
##vamos a ver realmente si fue por eso

df['Avg_Attendance'] = df['Attendance'] / df['MatchesPlayed']
df.sort_values('Avg_Attendance',ascending = False).head()


,Year,Country,Winner,Runners-Up,Third,Fourth,GoalsScored,QualifiedTeams,MatchesPlayed,Attendance,Avg_Attendance
14,1994,USA,Brazil,Italy,Sweden,Bulgaria,141,24,52,3587538,68991.115385
21,2022,Qatar,Argentina,France,Croatia,Morocco,172,32,64,3404252,53191.437500
19,2014,Brazil,Germany,Argentina,Netherlands,Brazil,171,32,64,3386810,52918.906250
17,2006,Germany,Italy,France,Germany,Portugal,147,32,64,3359439,52491.234375
8,1970,Mexico,Brazil,Italy,Germany FR,Uruguay,95,16,32,1603975,50124.218750


 USA 1994 lidera también en asistencia media por partido (~68.991), por delante de Mundiales más modernos con más partidos. Confirma que el récord se debe a los grandes estadios estadounidenses, no solo al número de encuentros.


**¿Qué país ha ganado más Mundiales en la historia?**

In [7]:
df['Winner'].value_counts()

,count
Winner,
Brazil,5
Italy,4
Germany FR,3
Argentina,3
Uruguay,2
France,2
England,1
Spain,1
Germany,1


In [8]:
#Lo de Alemania es un poco raro, habría que cambiarlo

df['Winner'] = df['Winner'].replace('Germany FR','Germany')
df['Winner'].value_counts()

,count
Winner,
Brazil,5
Italy,4
Germany,4
Argentina,3
Uruguay,2
France,2
England,1
Spain,1


 Brasil, con 5 títulos, seguido de Italia y Alemania con 4.(Nota: hubo que unificar "Germany FR" y "Germany", que el dataset separaba, para no infravalorar a Alemania).

In [9]:
#Vamos a guardar la  tabla de partidos en otro dataframe

df_matches = pd.read_csv('https://raw.githubusercontent.com/claudiocdc/fifa-world-cup-analysis/refs/heads/main/WorldCupMatches.csv')
print(df_matches.head())
print(df_matches.info())


     Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0

In [10]:
#attendance tiene menos, ¿Por qué? y hay muchas filas vacias ya que pone que hay 4572 entries cuando realmente hay solo rellenas 852
##Hay una función muy buena que es .dropna() https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html
###Segñun la docu el how ='all' permite borrar las que estan enteramente vacias y no solo las que tengan alguna vacía

df_matches = df_matches.dropna(how='all')
print(df_matches.info())


<class 'pandas.core.frame.DataFrame'>
Index: 852 entries, 0 to 851
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  852 non-null    float64
 1   Datetime              852 non-null    object 
 2   Stage                 852 non-null    object 
 3   Stadium               852 non-null    object 
 4   City                  852 non-null    object 
 5   Home Team Name        852 non-null    object 
 6   Home Team Goals       852 non-null    float64
 7   Away Team Goals       852 non-null    float64
 8   Away Team Name        852 non-null    object 
 9   Win conditions        852 non-null    object 
 10  Attendance            850 non-null    float64
 11  Half-time Home Goals  852 non-null    float64
 12  Half-time Away Goals  852 non-null    float64
 13  Referee               852 non-null    object 
 14  Assistant 1           852 non-null    object 
 15  Assistant 2           852 no

**¿Cuántos mundiales ha jugado cada país?**

In [11]:
#Hay que juntar las veces que aparece tanto local como visitante para ver sus apariciones, luego usar la variable de año como indicador para contar cuantas veces aparecieron
##Aparte year esta en float y queda feo, vamos a cambiarlo a int
###Al ejecutar el código más adelante, vemos que hay nombres en sucio con valores como rn">Bosnia and Herzegovina	porlo que quitamos ese 'rn">'

df_matches["Home Team Name"] = df_matches["Home Team Name"].str.replace('rn">', '')
df_matches["Away Team Name"] = df_matches["Away Team Name"].str.replace('rn">', '')
df_matches['Home Team Name'] = df_matches['Home Team Name'].replace('Germany FR','Germany')
df_matches['Away Team Name'] = df_matches['Away Team Name'].replace('Germany FR','Germany')

df_matches['Year'] = df_matches['Year'].astype(int)
home = df_matches[['Year','Home Team Name']].rename(columns={'Home Team Name':'Team'})
away = df_matches[['Year','Away Team Name']].rename(columns={'Away Team Name':'Team'})
appearances = pd.concat([home, away])
appearances

,Year,Team
0,1930,France
1,1930,USA
2,1930,Yugoslavia
3,1930,Romania
4,1930,Argentina
...,...,...
847,2014,Costa Rica
848,2014,Germany
849,2014,Argentina
850,2014,Netherlands


In [12]:
#Contamos las apariciones por año: agrupar por equipo, y de cada grupo cuántos años únicos tiene. Ojo con la n en el unique
##Confirmamos que Brazil es el que más mundiales ha ganado, también poruqe es el que más mundiales ha jugado

appearances.groupby('Team')['Year'].nunique().sort_values(ascending=False)



,Year
Team,
Brazil,20
Germany,18
Italy,18
Argentina,16
Mexico,15
...,...
Togo,1
United Arab Emirates,1
Ukraine,1


Brasil es la única selección que ha jugado las 20 ediciones (1930–2014). Le siguen Italia y Argentina.

**¿Cuántas veces apareció España?**

In [13]:
# Consulta rápida:

consult_country = 'Spain'
rows_spain = appearances[appearances['Team'] == consult_country]
World_cups = rows_spain['Year'].nunique()
print('El país', consult_country, 'ha jugado', World_cups, 'Mundiales hasta 2014')

El país Spain ha jugado 14 Mundiales hasta 2014


**¿Cuál ha sido el partido con más goles de toda la historia de los Mundiales?**

In [14]:
#Creamos columna con el total de goles y luego ordenamos
##Voy a pasar ambas columnas a int ya que no puedes marcar un gol y medio
###Creamos la variable y ordenamos

df_matches[['Away Team Goals','Home Team Goals']] = df_matches[['Away Team Goals','Home Team Goals']].astype(int)
df_matches['total_goals'] = df_matches['Away Team Goals']+ df_matches['Home Team Goals']
df_matches.sort_values('total_goals',ascending = False).head(1)



,Year,Datetime,Stage,Stadium,City,Home Team Name,Home Team Goals,Away Team Goals,Away Team Name,Win conditions,...,Half-time Home Goals,Half-time Away Goals,Referee,Assistant 1,Assistant 2,RoundID,MatchID,Home Team Initials,Away Team Initials,total_goals
94,1954,26 Jun 1954 - 17:00,Quarter-finals,La Pontaise,Lausanne,Austria,7,5,Switzerland,,...,5.0,4.0,FAULTLESS Charlie (SCO),ASENSI Manuel (ESP),SCHMETZER Emil (FRG),212.0,1237.0,AUT,SUI,12


Austria 7–5 Suiza (1954), con 12 goles.

**¿Porqué se metieron tantos goles? ¿Jugaron todo el día?**

In [15]:
#vamos a guardar los datos del partido para trabajarlos

top_match = df_matches.sort_values('total_goals',ascending = False).head(1)
top_match[['Year','Home Team Name','Away Team Name','Home Team Goals','Away Team Goals','Half-time Home Goals','Half-time Away Goals','Win conditions']]

,Year,Home Team Name,Away Team Name,Home Team Goals,Away Team Goals,Half-time Home Goals,Half-time Away Goals,Win conditions
94,1954,Austria,Switzerland,7,5,5.0,4.0,


Y lo más llamativo: se marcaron en los 90 minutos reglamentarios (sin prórroga), 9 de ellos antes del descanso.

**¿Qué selección ha metido más goles en toda la historia de los Mundiales?**

In [16]:
#Vamos a crear una tabla con los equipos y los goles tanto de local como de visitante
##Renombramos los nombres para unificarlos https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html#pandas.DataFrame.rename
###Concatenamos y creamos una tabla con ello
####Finalmente agrupamos los equipos por la suma de goles

df_home_goals = df_matches[['Home Team Name','Home Team Goals']].rename(columns={'Home Team Name':'Team', 'Home Team Goals':'Goals'})
df_away_goals = df_matches[['Away Team Name','Away Team Goals']].rename(columns={'Away Team Name':'Team', 'Away Team Goals':'Goals'})

df_top_goals = pd.concat([df_home_goals, df_away_goals])

df_top_goals.groupby('Team')['Goals'].sum().sort_values(ascending=False)

,Goals
Team,
Germany,235
Brazil,225
Argentina,133
Italy,128
France,108
...,...
China PR,0
Canada,0
Dutch East Indies,0


Alemania es la máxima goleadora de la historia de los Mundiales, por delante de Brasil (225 goles). (Ojo: este resultado solo emerge tras unificar "Germany FR" y "Germany"; sin esa limpieza, Brasil parecía el número uno).

**Análisis de jugadores**

In [17]:
df_players = pd.read_csv('https://raw.githubusercontent.com/claudiocdc/fifa-world-cup-analysis/refs/heads/main/WorldCupPlayers.csv')
print(df_players.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37784 entries, 0 to 37783
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   RoundID        37784 non-null  int64 
 1   MatchID        37784 non-null  int64 
 2   Team Initials  37784 non-null  object
 3   Coach Name     37784 non-null  object
 4   Line-up        37784 non-null  object
 5   Shirt Number   37784 non-null  int64 
 6   Player Name    37784 non-null  object
 7   Position       4143 non-null   object
 8   Event          9069 non-null   object
dtypes: int64(3), object(6)
memory usage: 2.6+ MB
None


Nota: un NaN no siempre es "dato sucio". A veces significa "no aplica" o "no ocurrió". En este caso es normal que haya jugadores que no registren eventos (goles,tarjetas...) y el position se entiende que porque se rellena si eres capitan por ejemplo


In [18]:
df_players['Position'].value_counts()

,count
Position,
GK,2441
C,1510
GKC,192


Vemos que Position registra marca los roles de portero (goalkeeper) y capitan (captain)

**¿Cuál es el número de camiseta más usado en la historia?**

In [19]:
df_players['Shirt Number'].value_counts()

,count
Shirt Number,
0,3069
1,1559
2,1554
3,1554
4,1554
7,1554
8,1554
9,1554
16,1554


Es un poco raro lo del 0, no sabemos si porque registra a los jugadores que no jugaron con el dorsal 0 o porque antiguamente no usaban dorsal.

In [20]:
#Sabemos que necesitamos que las dos columnas clave para unir dos dataframes tienen que tener mismo formato, sino da error. Igualamos el dtype de la columna de ambas tablas
df_matches['MatchID'] = df_matches['MatchID'].astype(int)

##Unimos por año de partidos a la tabla de jugadores
df_players_year = df_players.merge(df_matches[['MatchID','Year']], how='left', on='MatchID')
print(df_players_year.head())


   RoundID  MatchID Team Initials           Coach Name Line-up  Shirt Number  \
0      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
1      201     1096           MEX     LUQUE Juan (MEX)       S             0   
2      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
3      201     1096           MEX     LUQUE Juan (MEX)       S             0   
4      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   

        Player Name Position Event  Year  
0       Alex THEPOT       GK   NaN  1930  
1   Oscar BONFIGLIO       GK   NaN  1930  
2  Marcel LANGILLER      NaN  G40'  1930  
3      Juan CARRENO      NaN  G70'  1930  
4   Ernest LIBERATI      NaN   NaN  1930  


In [21]:
#Ahora vamos a ver cuantos 0 habian por año para ver si es la hipotesis de los mundiales sin dorsales u otra cosa
row_dorsal_0 = df_players_year[df_players_year['Shirt Number'] == 0]
row_dorsal_0['Year'].value_counts()



,count
Year,
1950,918
1938,762
1934,702
1930,687


Los dorsales no se usaron en los Mundiales hasta 1954; antes, los jugadores no llevaban número.

**¿Cuál sería el número más usado si quitamos los ceros?**

In [23]:
row_dorsal = df_players_year[df_players_year['Shirt Number'] != 0]
row_dorsal['Shirt Number'].value_counts()

,count
Shirt Number,
1,1623
2,1618
3,1618
4,1618
7,1618
8,1618
9,1618
16,1618
10,1618


Excluyendo el "0" (la era sin dorsales, hasta 1950), el número más usado en la historia es el 1, el dorsal del portero.